[v0.10.0 - > v0.13.0] Task9 (#223)
hyperai/tvm-cn: docs/how_to/deploy/09-deploy_adreno.md

> # 部署到Adreno™ GPU

Reviewer: ```suggestion
# 部署到 Adreno™ GPU
```

## 介绍

> Adreno™ 是由高通开发并用于许多SoC的图形处理单元（GPU）半导体IP核系列。

Reviewer: ```suggestion
Adreno™ 是由高通开发并用于许多 SoC 的图形处理单元（GPU）半导体 IP 核系列。
```

Adreno™ GPU 可以加速复杂几何图形的渲染，在提供高性能图形和丰富的用户体验的同时拥有很低的功耗。

TVM 使用 TVM 的原生 OpenCL 后端 和 OpenCLML 后端以支持加速 Adreno™ GPU 上的深度学习。TVM 的原生 OpenCL 后端通过结合纹理内存使用和 Adreno™ 友好布局来改进 Adreno™ 。 OpenCLML 是由高通发布的 SDK ，提供了大多数深度学习运算符的内核加速库。

本指南展示以下方面的不同设计

> * [OpenCL后端增强](#opencl-后端增强)

Reviewer: ```suggestion
* [OpenCL 后端增强](#opencl-后端增强)
```

* [关于 OpenCLML](#关于-OpenCLML)
* [构建与部署](#Adreno™-下的-TVM)

## OpenCL 后端增强

TVM 的 OpenCL 后端已被增强以利用 Adreno™ 特色功能，如
- 纹理内存使用。
- Adreno™ 友好的激活布局。
- 全新的调度以加速上述功能。

Adreno™ 的一个优势是对纹理的巧妙处理。目前，TVM 能够通过对 Adreno™  的纹理支持获得益处。下图显示了 Adreno™ A5x 架构。

> ![ Adreno™ A5x架构的高级概览，用于OpenCL](https://raw.githubusercontent.com/tlc-pack/web-data/main/images/how-to/adreno_architecture.png)

Reviewer: ```suggestion
![ Adreno™ A5x 架构的高级概览，用于 OpenCL](https://raw.githubusercontent.com/tlc-pack/web-data/main/images/how-to/adreno_architecture.png)
```

*图1 用于 OpenCL Adreno™ A5x 架构的高层次概览*

*来源：* [Qualcomm Adreno™ GPU 的 OpenCL 优化和最佳实践](https://dl.acm.org/doi/10.1145/3204919.3204935)

使用纹理的原因：

- 纹理处理器（TP）具有专用的 L1 缓存，它是只读缓存，并存储了从 2 级缓存（L2）中提取的用于纹理操作的数据（主要原因）

- 图像边界的处理已内置。

- 支持多种图像格式和数据类型组合，支持自动格式转换

总体而言，与基于 OpenCL 缓冲区的解决方案相比，使用纹理可以获得显著的性能提升。

通常，我们将目标指定为 ``target="opencl"`` ，以生成如下所示的内核的常规OpenCL目标。

```c
   __kernel void tvmgen_default_fused_nn_conv2d_kernel0(__global float* restrict p0, __global double* restrict p1, __global float* restrict conv2d_nhwc) {
   // body..

```

上述 OpenCL 内核定义有 ``__global float*`` 类型指针，它们实质上是 OpenCL ``buffer`` 对象。

通过修改目标定义为 ``target="opencl -device=adreno"`` 启用基于纹理的增强后，我们可以看到生成的内核使用纹理支持的 OpenCL 图像对象，如下所示。

```c
   __kernel void tvmgen_default_fused_nn_conv2d_kernel0(__write_only image2d_t pad_temp_global_texture, __read_only image2d_t p0) {
   // body..
```

*image2d_t* 是内置的 OpenCL 类型，用于表示二维图像对象并提供几个附加功能。
当我们使用 *image2d_t* 时，我们一次读取 *4个元素* 可以更有效地利用硬件。

> 有关生成和检查内核源的更多详细信息，请参阅[高级用法](高级用法)。

Reviewer: 这里的链接前面需要增加 `#` 否则会被认为是外部链接

```suggestion
有关生成和检查内核源的更多详细信息，请参阅[高级用法](#高级用法)。
```

## 关于 OpenCLML

OpenCLML 是高通发布的 SDK ，提供了更快的深度学习运算符。这些运算符作为标准 OpenCL 规范的扩展 ``cl_qcom_ml_ops`` 公开。有关更多详细信息，请参见 [Accelerate your models with our OpenCL ML SDK](https://developer.qualcomm.com/blog/accelerate-your-models-our-opencl-ml-sdk) 。

OpenCLML 已集成到 TVM ,作为 [BYOC](https://tvm.apache.org/docs/dev/how_to/relay_bring_your_own_codegen.html?highlight=bring%20your%20own) 解决方案。 OpenCLML 运算符可以使用相同的上下文，并可以加入到同样被原生 OpenCL 使用的命令队列中。我们利用了这一点避免在回退到原生 OpenCL 时的上下文切换。

## Adreno™ 下的 TVM

本节提供有关构建和部署模型到 Adreno™ 目标机的方法说明。 Adreno™ 是通过 ADB 连接与主机连接的远程目标。在这里部署已编译的模型需要在主机和目标上使用一些工具。

TVM 提供了简单的、用户友好的命令行工具以及专为开发者设计的 Python API 接口，可用于各种步骤，如自动调整、构建和部署。

![ 图2 ](https://raw.githubusercontent.com/tlc-pack/web-data/main/images/how-to/Adreno-Deployment-Pipeline.jpg)
*图2 Adreno 设备的构建和部署流水线*

上图展示了以下各个阶段的通用流程：

> * 导入模型： 在此阶段，我们从知名框架（如 TensorFlow 、 PyTorch 、 ONNX 等）导入模型。该阶段将给定的模型转换为 TVM 的 relay 模块格式。或者也可以通过使用 TVM 的操作库手动构建 relay 模块。此处生成的 TVM 模块的是独立于图的表示形式的目标。

Reviewer: ```suggestion
* 导入模型：在此阶段，我们从知名框架（如 TensorFlow、PyTorch、ONNX 等）导入模型。该阶段将给定的模型转换为 TVM 的 relay 模块格式。或者也可以通过使用 TVM 的操作库手动构建 relay 模块。此处生成的 TVM 模块的是独立于图的表示形式的目标。
```

* 自动调整：在此阶段，我们调整特定于目标的 TVM 生成的内核。自动调整过程需要目标设备的有效性，在如 Android 设备上的 Adreno™ 这样的远程目标中，我们使用 RPC 设置进行通信。本指南的后续部分将详细介绍 Android 设备的 RPC 设置。自动调整不是模型编译的必需步骤，但对于获得 TVM 生成的内核的最佳性能是必要的。

* 编译：在此阶段，我们为特定目标编译模型。鉴于我们在前一阶段自动调整了模块， TVM 编译利用调整日志以生成性能最佳的内核。 TVM 编译进程中产生包含内核的共享库、以 json 格式定义的图和以 TVM 特定格式的二进制参数文件。

* 部署（或测试运行）到目标：在此阶段，我们在目标上运行 TVM 编译输出。部署可以从 Python 环境中使用 RPC 设置。也可以使用 TVM 的本地工具进行，该工具可以为 Android 进行本地二进制交叉编译。在此阶段，我们可以在 Android 目标上运行已编译的模型，并对输出的正确性和性能方面进行单元测试。

* 应用集成：本阶段涉及将 TVM 编译的模型集成到应用程序中。我们在这里讨论如何从 Android（cpp 本地环境或 JNI）中设置输入和获取输出的 TVM 运行时接口。 

* 高级用法：本部分涵盖了高级用户感兴趣的主题，如查看生成的源代码、更改模块的精度等。

此教程以下各节将涵盖上述内容。

- [开发环境](#自动开发环境设置)
- [RPC 设置](#rpc-设置)
- [命令行工具](#命令行工具)
- [Python 接口](#python-接口)
- [应用集成](#应用程序集成)
- [高级用法](#高级用法)

## 自动开发环境设置

> TVM 提供了一个预定义的 Docker 容器环境，其中包含了所有入门所需的先决条件。如果您想要更多地控制依赖关系，请参考[手动环境设置](手动环境设置)。

Reviewer: 这里的链接前面需要增加 `#` 否则会被认为是外部链接

```suggestion
TVM 提供了一个预定义的 Docker 容器环境，其中包含了所有入门所需的先决条件。如果您想要更多地控制依赖关系，请参考[手动环境设置](#手动环境设置)。
```

对于 Docker 设置，先决条件只是主机上有 Docker 工具。

以下命令可以构建 Adreno 的 Docker 镜像：

```bash
./docker/build.sh ci_adreno
docker tag tvm.ci_adreno ci_adreno
```

现在，我们可以使用以下命令构建主机和目标工具：

```bash
./tests/scripts/ci.py adreno -i
```

> 要使用 OpenCLML SDK 构建TVM，在构建时需要导出 OpenCLML SDK ，如下所示：

Reviewer: ```suggestion
要使用 OpenCLML SDK 构建TVM，在构建时需要导出 OpenCLML SDK，如下所示：
```

```bash
export ADRENO_OPENCL=<OpenCLML SDK路径>
./tests/scripts/ci.py adreno -i
```

> 成功编译后，您将进入Docker shell。构建将生成两个文件夹：

Reviewer: ```suggestion
成功编译后，您将进入 Docker shell。构建将生成两个文件夹：
```

> - `build-adreno` ：主机端TVM编译器构建。
> - `build-adreno-target` ：包含 Android 目标组件。
>   - libtvm_runtime.so : TVM 运行时库
>   - tvm_rpc : rpc 运行时环境工具
>   - rtvm : 独立的原生工具

Reviewer: 这里统一使用中文全角冒号：

```suggestion
- `build-adreno`：主机端TVM编译器构建。
- `build-adreno-target`：包含 Android 目标组件。
  - libtvm_runtime.so：TVM 运行时库
  - tvm_rpc：rpc 运行时环境工具
  - rtvm：独立的原生工具
```

在使用Docker环境时， Android 设备与主机共享，所以在主机上需要安装 ADB 版本为 ``1.0.41`` ，因为 Docker 使用相同的版本。

您还可以在 Docker 环境中检查 ADB 设备的可用性：

```bash
user@ci-adreno-fpeqs:~$ adb devices
List of devices attached
aaaabbbb	device
ccccdddd	device
```

## 手动开发环境设置

手动构建过程需要构建主机和目标组件。

以下命令将配置主机编译：

```bash
mkdir -p build
cd build
cp ../cmake/config.cmake .

> # 启用RPC功能以与远程设备通信。
> echo set\(USE_RPC ON\) >> config.cmake
> # 我们在主机（x86）上使用图执行器验证模型。
> echo set\(USE_GRAPH_EXECUTOR ON\) >> config.cmake
> # 启用溢出时的回溯以获取更多调试信息。
> echo set\(USE_LIBBACKTRACE AUTO\) >> config.cmake
> # 目标主机将是llvm。
> echo set\(USE_LLVM ON\) >> config.cmake

Reviewer: ```suggestion
# 启用 RPC 功能以与远程设备通信。
echo set\(USE_RPC ON\) >> config.cmake
# 我们在主机（x86）上使用图执行器验证模型。
echo set\(USE_GRAPH_EXECUTOR ON\) >> config.cmake
# 启用溢出时的回溯以获取更多调试信息。
echo set\(USE_LIBBACKTRACE AUTO\) >> config.cmake
# 目标主机将是 llvm。
echo set\(USE_LLVM ON\) >> config.cmake
```

```

此外，我们可以推送以下配置条目，以使用 OpenCLML 支持进行编译：

```bash
export ADRENO_OPENCL=<OpenCLML SDK路径>
echo set\(USE_CLML ${ADRENO_OPENCL}\) >> config.cmake
```

现在我们可以像下面这样构建：

```bash
cmake ..
make
```

> 最后，我们可以导出Python路径：

Reviewer: ```suggestion
最后，我们可以导出 Python 路径：
```

```bash
export PYTHONPATH=$TVM_HOME/python:${PYTHONPATH}
python3 -c "import tvm" # 验证tvm Python包
```

> 现在，我们可以使用以下配置来配置和构建目标组件。目标构建需要安装Android NDK。

Reviewer: ```suggestion
现在，我们可以使用以下配置来配置和构建目标组件。目标构建需要安装 Android NDK。
```

> - 在此处阅读有关 *Android NDK安装* 的文档：[https://developer.android.com/ndk](https://developer.android.com/ndk)
> - 要获取adb工具的访问权限，您可以在此处查看 *Android Debug Bridge安装* ：[https://developer.android.com/studio/command-line/adb](https://developer.android.com/studio/command-line/adb)

Reviewer: ```suggestion
- 在此处阅读有关 *Android NDK 安装* 的文档：[https://developer.android.com/ndk](https://developer.android.com/ndk)
- 要获取 adb 工具的访问权限，您可以在此处查看 *Android Debug Bridge 安装* ：[https://developer.android.com/studio/command-line/adb](https://developer.android.com/studio/command-line/adb)
```

```bash
mkdir -p build-adreno
cd build-adreno
cp ../cmake/config.cmake .
# 启用 OpenCL 后端。
echo set\(USE_OPENCL ON\) >> config.cmake
# 启用 RPC 功能。
echo set\(USE_RPC ON\) >> config.cmake
# 构建在目标设备上运行的 tvm_rpc 工具。
echo set\(USE_CPP_RPC ON\) >> config.cmake
# 构建本机 rtvm 部署工具。
echo set\(USE_CPP_RTVM ON\) >> config.cmake
# 我们在像 Android 这样的设备上使用图执行器。
echo set\(USE_GRAPH_EXECUTOR ON\) >> config.cmake
# 在可能的情况下启用回溯。
echo set\(USE_LIBBACKTRACE AUTO\) >> config.cmake
# Adreno 支持 OpenCL 分配的32位对齐而不是64位。
echo set\(USE_KALLOC_ALIGNMENT 32\) >> config.cmake

# Android 构建相关定义。
echo set\(ANDROID_ABI arm64-v8a\) >> config.cmake
echo set\(ANDROID_PLATFORM android-28\) >> config.cmake
echo set\(MACHINE_NAME aarch64-linux-gnu\) >> config.cmake
```

此外，我们可以推送以下配置以使用 OpenCLML 支持进行编译：

```bash
export ADRENO_OPENCL=<OpenCLML SDK路径>
echo set\(USE_CLML "${ADRENO_OPENCL}"\) >> config.cmake
echo set\(USE_CLML_GRAPH_EXECUTOR "${ADRENO_OPENCL}"\) >> config.cmake
```

> 对于Android目标构建， ``ANDROID_NDK_HOME`` 是一个依赖项，我们应该在环境变量中设置相同的依赖项。以下命令将构建 Adreno™ 目标组件：

Reviewer: ```suggestion
对于 Android 目标构建，``ANDROID_NDK_HOME`` 是一个依赖项，我们应该在环境变量中设置相同的依赖项。以下命令将构建 Adreno™ 目标组件：
```

cmake -DCMAKE_TOOLCHAIN_FILE="${ANDROID_NDK_HOME}/build/cmake/android.toolchain.cmake" \
   -DANDROID_ABI=arm64-v8a \
   -DANDROID_PLATFORM=android-28 \
   -DCMAKE_SYSTEM_VERSION=1 \
   -DCMAKE_FIND_ROOT_PATH="${ADRENO_OPENCL}" \
   -DCMAKE_FIND_ROOT_PATH_MODE_PROGRAM=NEVER \
   -DCMAKE_FIND_ROOT_PATH_MODE_LIBRARY=ONLY \
   -DCMAKE_CXX_COMPILER="${ANDROID_NDK_HOME}/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android28-clang++" \
   -DCMAKE_C_COMPILER="${ANDROID_NDK_HOME}/toolchains/llvm/prebuilt/linux-x86_64/bin/aarch64-linux-android28-clang" \
   -DMACHINE_NAME="aarch64-linux-gnu" ..

make tvm_runtime tvm_rpc rtvm
```

## RPC 设置

RPC 设置允许通过 TCP/IP 网络接口访问远程目标。RPC 设置对于自动调整阶段是必不可少的，因为调整涉及在真实设备上运行自动生成的内核，并通过使用机器学习方法对其进行优化。请参考 [使用模板和 AutoTVM 进行自动调整](https://tvm.apache.org/docs/how_to/tune_with_autotvm/index.html) 以获取有关 AutoTVM 的详细信息。

RPC 设置也可用于通过 python 接口或来自主机设备的 ``tvmc`` 工具将已编译的模型部署到远程设备。

RPC 设置有多个组件，如下所示。

**TVM Tracker:** TVM Tracker 是主机端的守护程序，管理远程设备并为主机端应用程序提供服务。应用程序可以连接到此跟踪器并获取远程设备句柄以进行通信。

**TVM RPC:** TVM RPC 是在远程设备上运行的本地应用程序（本情况下是 Android ），并向在主机上运行的 TVM Tracker 注册自身。

因此，对于基于 RPC 的设置，我们将在主机和目标设备上运行上述组件。以下部分解释了如何手动设置相同的内容，以及如何在使用自动化工具的 Docker 环境中设置相同的内容。

> **自动化RPC设置:**此处，我们将解释如何在 Docker 环境中设置 RPC 。

Reviewer: ```suggestion
**自动化 RPC 设置:**此处，我们将解释如何在 Docker 环境中设置 RPC 。
```

> 以下命令在 Docker 环境中启动 tracker，其中 tracker 监听端口9190。

Reviewer: ```suggestion
以下命令在 Docker 环境中启动 tracker，其中 tracker 监听端口 9190。
```

```bash
./tests/scripts/ci.py adreno -i # 在 anreno docker 上启动一个新的 shell
source tests/scripts/setup-adreno-env.sh -e tracker -p 9190
```

现在，以下命令可以在具有 ID ``abcdefgh`` 的远程 Android 设备上运行 TVM RPC 。

```bash
./tests/scripts/ci.py adreno -i # 在adreno docker上启动一个新的shell。
source tests/scripts/setup-adreno-env.sh -e device -p 9190 -d abcdefgh
```

此外，以下命令可用于在任何其他 Docker 终端上查询 RPC 设置详细信息。

```bash
./tests/scripts/ci.py adreno -i # 在 adreno docker 上启动一个新的 shell。
source tests/scripts/setup-adreno-env.sh -e query -p 9190
```

> **手动RPC设置:**请参阅教程 [在 Adreno 上部署模型](https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno.html)了解手动 RPC 环境设置。

Reviewer: ```suggestion
**手动 RPC 设置：**请参阅教程 [在 Adreno 上部署模型](https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno.html) 了解手动 RPC 环境设置。
```

> 这些 RPC 设置完成后，我们在主机 ``127.0.0.1``（rpc-tracker）上拥有rpc-tracker，端口 ``9190`` （rpc-port）可用。

Reviewer: ```suggestion
这些 RPC 设置完成后，我们在主机 ``127.0.0.1``（rpc-tracker）上拥有 rpc-tracker，端口 ``9190`` （rpc-port）可用。
```

.. _commandline_interface:

## 命令行工具

> 此处我们使用命令行工具描述整个编译进程。 TVM 具有命令行实用程序 [tvmc](https://tvm.apache.org/docs/tutorial/tvmc_command_line_driver.html) ，用于执行模型导入、自动调整、编译和rpc部署。 [tvmc](https://tvm.apache.org/docs/tutorial/tvmc_command_line_driver.html) 有许多选项可供探索和尝试。

Reviewer: ```suggestion
此处我们使用命令行工具描述整个编译进程。 TVM 具有命令行实用程序 [tvmc](https://tvm.apache.org/docs/tutorial/tvmc_command_line_driver.html)，用于执行模型导入、自动调整、编译和 rpc 部署。 [tvmc](https://tvm.apache.org/docs/tutorial/tvmc_command_line_driver.html) 有许多选项可供探索和尝试。
```

> **模型导入和调整:**使用以下命令从任何框架导入模型并对其进行自动调整。此处我们使用来自 Keras 的模型，它使用 RPC 设置进行调整，并最终生成调整日志文件 ``keras-resnet50.log`` 。

Reviewer: ```suggestion
**模型导入和调整：**使用以下命令从任何框架导入模型并对其进行自动调整。此处我们使用来自 Keras 的模型，它使用 RPC 设置进行调整，并最终生成调整日志文件 ``keras-resnet50.log`` 。
```

在将其集成到现有 Android 应用程序中时， TVM 有多种选择。对于 JNI 或 CPP 原生，我们可以使用 [C Runtime API](https://github.com/apache/tvm/blob/main/include/tvm/runtime/c_runtime_api.h) 。
您还可以参考 ``rtvm`` 的简化接口 [TVMRunner](https://github.com/apache/tvm/blob/main/apps/cpp_rtvm/tvm_runner.h>) 。

## Python 接口

该节解释如何使用 Python 接口导入、自动调整、编译和运行模型的过程。 TVM 通过 ``tvmc`` 抽象提供高级接口，以及低级的 Relay API 。我们将详细讨论这两者。

**TVMC 接口：**

在使用 ``tvmc`` Python接口时，我们首先加载一个生成 ``TVMCModel`` 的模型。 ``TVMCModel`` 将用于自动调整以生成调整缓存。编译进程使用 ``TVMCModel`` 和调整缓存（可选）来生成 ``TVMCPackage`` 。现在 ``TVMCPackage`` 将保存到文件系统，或者可以用于部署和在目标设备上运行。

请参考相应的教程[如何使用 TVMC 在 Adreno 上部署预训练模型](https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno_tvmc.html)。

保存的 ``TVMCPackage`` 也可以用于使用 ``rtvm`` 工具进行原生部署。

此外，请参考 [使用 TVMC 编译和优化模型](https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno.html) 文档，了解有关 API 接口的更多详细信息。

**Relay 接口：**

Relay API 接口提供了对 TVM 编译器接口的低级 API 访问。与 ``tvmc`` 接口类似，Relay API 接口提供了各种前端 API，用于将模型转换为 Relay ``Module``。Relay ``Module`` 将用于所有种类的转换，如精度转换、CLML 卸载和其他自定义转换（如果有）。生成的 Module 也将用于自动调整。最后，我们使用 ``relay.build`` API 生成库模块。从这个库模块，我们可以导出编译产物，如模块共享库（mod.so）、参数（mod.params）和 json 图（mod.json）。这个库模块将用于创建图形运行时以在目标设备上部署和运行。

> 请参考教程[如何在 Adreno 上部署预训练模型](https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno.html),其中逐步解释了相同的内容。

Reviewer: ```suggestion
请参考教程[如何在 Adreno 上部署预训练模型](https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno.html)，其中逐步解释了相同的内容。
```
